### PR TITLE
Adding an example with truststore to bypass cert pinning

### PR DESCRIPTION
**Pull request by lwierzbicki**
This PR covers issue #1612 .

### CODE REVIEW

File: Document/0x05g-Testing-Network-Communication.md [hunk 1]
```diff
@@ -338,6 +338,7 @@ Dynamic analysis can be performed by launching a MITM attack with your preferred
 
 There are several ways to bypass certificate pinning for a black box test, depending on the frameworks available on the device:
 
+- Frida: Use the [Universal Android SSL Pinning Bypass with Frida](https://codeshare.frida.re/@pcipolloni/universal-android-ssl-pinning-bypass-with-frida/ "Universal Android SSL Pinning Bypass with Frida") script.
 - Objection: Use the `android sslpinning disable` command.
 - Xposed: Install the [TrustMeAlready](https://github.com/ViRb3/TrustMeAlready "TrustMeAlready") or [SSLUnpinning](https://github.com/ac-pm/SSLUnpinning_Xposed "SSLUnpinning") module.
 - Cydia Substrate: Install the [Android-SSL-TrustKiller](https://github.com/iSECPartners/Android-SSL-TrustKiller "Android-SSL-TrustKiller") package.
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]: Could you order the list according to the alphabet and make "Cydia.." the first item?

File: Document/0x05g-Testing-Network-Communication.md [hunk 2]
```diff
@@ -350,11 +351,32 @@ Somewhere in the application, both the endpoint and the certificate (or its hash
...
 - Certificate files: `find ./assets -type f \( -iname \*.cer -o -iname \*.crt \)`. Replace these files with your proxy's certificates, making sure they are in the correct format.
+- Truststore files: `find ./ -type f \( -iname \*.jks -o -iname \*.bks \)`. Add your proxy's certificates to trustore, making sure they are in the correct format.
 
-If the application uses native libraries to implement network communication, further reverse engineering is needed. An example of such an approach can be found in the blog post [Identifying the SSL Pinning logic in smali code, patching it, and reassembling the APK](https://serializethoughts.wordpress.com/2016/08/18/bypassing-ssl-pinning-in-android-applications/ "Bypassing SSL Pinning in Android Applications")
+> Consider situation when app contains files without extension. Most common files location are `assets` and `res` directories.
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]:
```suggestion
> Keep in mind that an app might contain files without extension. The most common file locations are `assets` and `res` directories, which should also be investigated.
```

File: Document/0x05g-Testing-Network-Communication.md [hunk 3]
```diff
@@ -350,11 +351,32 @@ Somewhere in the application, both the endpoint and the certificate (or its hash
...
+
+As an example, let's say that you find an application which uses a BKS (BouncyCastle) truststore and it's stored in file `res/raw/truststore.bks`. To bypass SSL Pinning you need to add your proxy's certificate. Following values need to be defined:
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]:
```suggestion
As an example, let's say that you find an application which uses a BKS (BouncyCastle) truststore and it's stored in the file `res/raw/truststore.bks`. To bypass SSL Pinning you need to add your proxy's certificate to the truststore with the command line tool `keytool`. `Keytool` comes with the Java SDK and the following values are needed to execute the command:
```

File: Document/0x05g-Testing-Network-Communication.md [hunk 4]
```diff
@@ -350,11 +351,32 @@ Somewhere in the application, both the endpoint and the certificate (or its hash
...
+- password - password to keystore, you can find it in app code, where keystore is open
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]:
```suggestion
- password - Password for the keystore. Look in the decompiled app code for the hardcoded password.
```

File: Document/0x05g-Testing-Network-Communication.md [hunk 5]
```diff
@@ -350,11 +351,32 @@ Somewhere in the application, both the endpoint and the certificate (or its hash
...
+- providerpath - BouncyCastle Provider jar location, you can download it from [The Legion of the Bouncy Castle](https://www.bouncycastle.org/latest_releases.html "https://www.bouncycastle.org/latest_releases.html")
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]:
```suggestion
- providerpath - Location of the BouncyCastle Provider jar file. You can download it from [The Legion of the Bouncy Castle](https://www.bouncycastle.org/latest_releases.html "https://www.bouncycastle.org/latest_releases.html").
```

File: Document/0x05g-Testing-Network-Communication.md [hunk 6]
```diff
@@ -350,11 +351,32 @@ Somewhere in the application, both the endpoint and the certificate (or its hash
...
+- proxy.cer - your proxy's certificate
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]:
```suggestion
- proxy.cer - Your proxy's certificate.
```

File: Document/0x05g-Testing-Network-Communication.md [hunk 7]
```diff
@@ -350,11 +351,32 @@ Somewhere in the application, both the endpoint and the certificate (or its hash
...
+- aliascert - unique value which will be used as alias for your proxy's certificate
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]:
```suggestion
- aliascert - Unique value which will be used as alias for your proxy's certificate.
```

File: Document/0x05g-Testing-Network-Communication.md [hunk 8]
```diff
@@ -350,11 +351,32 @@ Somewhere in the application, both the endpoint and the certificate (or its hash
...
+
+To add your proxy's certificate use the following command:
+
+```shell
+keytool -importcert -v -trustcacerts -file proxy.cer -alias aliascert -keystore "res/raw/truststore.bks" -provider org.bouncycastle.jce.provider.BouncyCastleProvider -providerpath "providerpath/bcprov-jdk15on-164.jar" -storetype BKS -storepass password
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]:
```suggestion
$ keytool -importcert -v -trustcacerts -file proxy.cer -alias aliascert -keystore "res/raw/truststore.bks" -provider org.bouncycastle.jce.provider.BouncyCastleProvider -providerpath "providerpath/bcprov-jdk15on-164.jar" -storetype BKS -storepass password
```

File: Document/0x05g-Testing-Network-Communication.md [hunk 9]
```diff
@@ -350,11 +351,32 @@ Somewhere in the application, both the endpoint and the certificate (or its hash
...
+```
+
+To list certificates in BKS truststore use the following command:
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]:
```suggestion
To list certificates in the BKS truststore use the following command:
```

File: Document/0x05g-Testing-Network-Communication.md [hunk 10]
```diff
@@ -350,11 +351,32 @@ Somewhere in the application, both the endpoint and the certificate (or its hash
...
+
+```shell
+keytool -list -keystore "res/raw/truststore.bks" -provider org.bouncycastle.jce.provider.BouncyCastleProvider -providerpath "providerpath/bcprov-jdk15on-164.jar"  -storetype BKS -storepass password
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]:
```suggestion
$ keytool -list -keystore "res/raw/truststore.bks" -provider org.bouncycastle.jce.provider.BouncyCastleProvider -providerpath "providerpath/bcprov-jdk15on-164.jar"  -storetype BKS -storepass password
```

File: Document/0x05g-Testing-Network-Communication.md (hunk 1 above)
Comment (lwierzbicki) [reply Document/0x05g-Testing-Network-Communication.md]: Done.

**Comment by lwierzbicki** [comment]
Last commit contain changes covers issue #1615

File: Document/0x04e-Testing-Authentication-and-Session-Management.md [hunk 11]
```diff
@@ -598,13 +598,13 @@ Note: in case of an application which requires L2 protection, it can be a good i
...
+- [grassi] Grassi, Paul A., et al. Digital identity guidelines: Authentication and lifecycle management (DRAFT). No. Special Publication (NIST SP)-800-63B. 2016.
+- [grassi2] Grassi, Paul A., et al. Digital identity guidelines: Authentication and lifecycle management. No. Special Publication (NIST SP)-800-63B. 2017.
+- [konoth] Konoth, Radhesh Krishnan, Victor van der Veen, and Herbert Bos. "How anywhere computing just killed your phone-based two-factor authentication." International Conference on Financial Cryptography and Data Security. Springer, Berlin, Heidelberg, 2016.
+- [mulliner] Mulliner, Collin, et al. "SMS-based one-time passwords: attacks and defense." International Conference on Detection of Intrusions and Malware, and Vulnerability Assessment. Springer, Berlin, Heidelberg, 2013.
+- [siadati] Siadati, Hossein, et al. "Mind your SMSes: Mitigating social engineering in second factor authentication." Computers & Security 65 (2017): 14-28.
+- [siadati2] Siadati, Hossein, Toan Nguyen, and Nasir Memon. "Verification code forwarding attack (short paper)." International Conference on Passwords. Springer, Cham, 2015.
```
Comment (sushi2k) [inline Document/0x04e-Testing-Authentication-and-Session-Management.md]: Could you add the "#" to the references in the brackets?

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 12]
```diff
@@ -2450,10 +2450,10 @@ File-hiding is of course only the tip of the iceberg: you can accomplish a lot u
 
 - Bionic - <https://github.com/android/platform_bionic>
 - Attacking Android Applications with Debuggers - <https://blog.netspi.com/attacking-android-applications-with-debuggers/>
-- Dynamic Malware Recompilation - <http://ieeexplore.ieee.org/document/6759227/>
+- [josse] Sébastien Josse, Dynamic Malware Recompilation - <http://ieeexplore.ieee.org/document/6759227/>
```
Comment (sushi2k) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: could you please add the hash sign?

File: Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 13]
```diff
@@ -670,7 +670,7 @@ Any scheme based on these methods will be more secure the moment a passcode and/
 
 ### References
 
-- Dana Geist, Marat Nigmatullin: Jailbreak/Root Detection Evasion Study on iOS and Android - <http://delaat.net/rp/2015-2016/p51/report.pdf>
+- [geist] Dana Geist, Marat Nigmatullin: Jailbreak/Root Detection Evasion Study on iOS and Android - <http://delaat.net/rp/2015-2016/p51/report.pdf>
```
Comment (sushi2k) [inline Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md]: Hash # missing

File: Document/0x06d-Testing-Data-Storage.md [hunk 14]
```diff
@@ -955,4 +955,4 @@ For more information, options and approaches, please refer to section "[In-Memor
 #### Others
 
 - Appthority Mobile Threat Team Research Paper - <https://cdn2.hubspot.net/hubfs/436053/Appthority%20Q2-2018%20MTR%20Unsecured%20Firebase%20Databases.pdf>
-- Demystifying the Secure Enclave Processor - <https://www.blackhat.com/docs/us-16/materials/us-16-Mandt-Demystifying-The-Secure-Enclave-Processor.pdf>
+- [mandt] Tarjei Mandt, Mathew Solnik  and  David Wang, Demystifying the Secure Enclave Processor - <https://www.blackhat.com/docs/us-16/materials/us-16-Mandt-Demystifying-The-Secure-Enclave-Processor.pdf>
```
Comment (sushi2k) [inline Document/0x06d-Testing-Data-Storage.md]: Hash # missing

File: Document/0x06h-Testing-Platform-Interaction.md [hunk 15]
```diff
@@ -3125,7 +3125,7 @@ Finally, see if you can play with the version number of a man-in-the-middled app
 
 ### References
 
-- [#THIEL] Thiel, David. iOS Application Security: The Definitive Guide for Hackers and Developers (Kindle Locations 3394-3399). No Starch Press. Kindle Edition.
+- [#THIEL] David Thiel, iOS Application Security: The Definitive Guide for Hackers and Developers (Kindle Locations 3394-3399), No Starch Press, Kindle Edition.
```
Comment (sushi2k) [inline Document/0x06h-Testing-Platform-Interaction.md]: please make THIEL lowercase

File: Document/0x05g-Testing-Network-Communication.md [hunk 16]
```diff
@@ -350,11 +351,32 @@ Somewhere in the application, both the endpoint and the certificate (or its hash
 
 - Certificate hashes: `grep -ri "sha256\|sha1" ./smali`. Replace the identified hashes with the hash of your proxy's CA. Alternatively, if the hash is accompanied by a domain name, you can try modifying the domain name to a non-existing domain so that the original domain is not pinned. This works well on obfuscated OkHTTP implementations.
 - Certificate files: `find ./assets -type f \( -iname \*.cer -o -iname \*.crt \)`. Replace these files with your proxy's certificates, making sure they are in the correct format.
+- Truststore files: `find ./ -type f \( -iname \*.jks -o -iname \*.bks \)`. Add your proxy's certificates to trustore, making sure they are in the correct format.
```
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]:
```suggestion
- Truststore files: `find ./ -type f \( -iname \*.jks -o -iname \*.bks \)`. Add your proxy's certificates to the trustore and make sure they are in the correct format.
```

File: Document/0x06h-Testing-Platform-Interaction.md (hunk 15 above)
Comment (lwierzbicki) [reply Document/0x06h-Testing-Platform-Interaction.md]: Done. Added as thiel2 as it is second refernce.

**Comment by lwierzbicki** [comment]
Last commit contains changes which covers issue #1618

**Pull request merged**